src/rulenode.jl

```diff
@@ -1,27 +1,41 @@
+"""
+Type for representing expression trees.
+"""
+abstract type AbstractRuleNode end
+
 
 """
 RuleNode
 Type for representing nodes in an expression tree.
 """
-mutable struct RuleNode
+mutable struct RuleNode <: AbstractRuleNode
 	ind::Int # index in grammar
 	_val::Any  #value of _() evals
-	children::Vector{RuleNode}
+	children::Vector{AbstractRuleNode}
+end
+
+mutable struct Hole <: AbstractRuleNode
+	domain::BitVector
 end
 
 RuleNode(ind::Int) = RuleNode(ind, nothing, RuleNode[])
+RuleNode(ind::Int, children::Vector{AbstractRuleNode}) = RuleNode(ind, nothing, children)
 RuleNode(ind::Int, children::Vector{RuleNode}) = RuleNode(ind, nothing, children)
+RuleNode(ind::Int, children::Vector{Hole}) = RuleNode(ind, nothing, children)
 RuleNode(ind::Int, _val::Any) = RuleNode(ind, _val, RuleNode[])
 
 include("recycler.jl")
 
-
-function Base.:(==)(A::RuleNode, B::RuleNode)
+function Base.:(==)(A::AbstractRuleNode, B::AbstractRuleNode)
```

Comment:
This is a small thing that can be skipped right now: it might be a good idea to try to leverage multiple dispatch in Julia as much as possible. So, instead of one big `Base.(==)` function, use four specific one:
`Base.(==)(::RuleNode, ::RuleNode)`
`Base.(==)(::RuleNode, ::Hole)`
`Base.(==)(::Hole, ::RuleNode)`
`Base.(==)(::Hole, ::Hole)`

I find this actually easier to read and I think Julia leverages this to compile the code even faster.

Reply:
Good suggestion, thanks!